File: paleo_workbench/ui/pages/qc_issue_table.py
from __future__ import annotations

from PySide6.QtCore import Qt
from PySide6.QtGui import QColor
from PySide6.QtWidgets import QHeaderView, QTableWidget, QTableWidgetItem, QVBoxLayout, QWidget

from paleo_workbench.ui import tokens
from paleo_workbench.ui.pages.qc_helpers import derive_rule_result

COLUMN_HEADERS = ["检查项目", "检查说明", "结果说明", "定位"]
COLUMN_WIDTHS = [160, 0, 160, 100]  # 0 = stretch (检查说明)


class QCIssueTable(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setObjectName("QCIssueTable")
        layout = QVBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        self.table = QTableWidget()
        self.table.setColumnCount(4)
        self.table.setHorizontalHeaderLabels(COLUMN_HEADERS)
        self.table.setAlternatingRowColors(True)
        # Inherit global QTableWidget / QHeaderView rules from tokens.QSS_TEMPLATE.
        self.table.verticalHeader().setVisible(False)
        self.table.setEditTriggers(QTableWidget.EditTrigger.NoEditTriggers)
        self.table.setSelectionBehavior(QTableWidget.SelectionBehavior.SelectRows)
        header = self.table.horizontalHeader()
        for i, w in enumerate(COLUMN_WIDTHS):
            if w > 0:
                header.resizeSection(i, w)
            else:
                header.setSectionResizeMode(i, QHeaderView.ResizeMode.Stretch)
        layout.addWidget(self.table)
        self._spatial_by_rule: dict[str, list[dict]] = {}

    def update_state(self, reports: list) -> None:
        self.table.setRowCount(0)
        self._spatial_by_rule = {}
        if not reports:
            return
        report = reports[0]
        from paleo_workbench.workflow.qc import spatial_issues

        for issue in spatial_issues(getattr(report, "issues", None)):
            rule = str(issue.get("rule") or "")
            self._spatial_by_rule.setdefault(rule, []).append(issue)

        for row, rule in enumerate(report.rules):
            self.table.insertRow(row)
            description = tokens.RULE_DESCRIPTIONS.get(rule, rule)
            severity, result_text, result_color = derive_rule_result(rule, report.issues)
            spatial = self._spatial_by_rule.get(rule) or []
            if spatial:
                first = spatial[0]
                loc = first.get("feature_id") or first.get("ref") or "可定位"
                if len(spatial) > 1:
                    loc = f"{loc} (+{len(spatial) - 1})"
            else:
                loc = "—"
            items = [
                QTableWidgetItem(rule),
                QTableWidgetItem(description),
                QTableWidgetItem(result_text),
                QTableWidgetItem(str(loc)),
            ]
            for col, item in enumerate(items):
                if col == 2:
                    item.setForeground(QColor(result_color))
                item.setFlags(item.flags() & ~Qt.ItemFlag.ItemIsEditable)
                self.table.setItem(row, col, item)
            self.table.setRowHeight(row, 28)

    def spatial_issues_for_rule(self, rule: str) -> list[dict]:
        return list(self._spatial_by_rule.get(rule) or [])

File: libs/ui_review/src/qt/qc_issue_table.cpp
#include "pwb/ui_review/qt/qc_issue_table.hpp"

#include "pwb/ui_review/qc_issue_rows.hpp"

#include <QColor>
#include <QHeaderView>
#include <QTableWidgetItem>
#include <QVBoxLayout>

namespace pwb::ui_review::qt {

QcIssueTable::QcIssueTable(QWidget* parent) : QWidget(parent) {
    setObjectName(QStringLiteral("QCIssueTable"));
    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(0, 0, 0, 0);
    table_ = new QTableWidget(this);
    table_->setColumnCount(4);
    // COLUMN_HEADERS verbatim.
    table_->setHorizontalHeaderLabels(
        {QStringLiteral("检查项目"), QStringLiteral("检查说明"),
         QStringLiteral("结果说明"), QStringLiteral("定位")});
    table_->setAlternatingRowColors(true);
    table_->verticalHeader()->setVisible(false);
    table_->setEditTriggers(QTableWidget::NoEditTriggers);
    table_->setSelectionBehavior(QTableWidget::SelectRows);
    // COLUMN_WIDTHS = [160, 0(stretch), 160, 100].
    auto* header = table_->horizontalHeader();
    header->resizeSection(0, 160);
    header->setSectionResizeMode(1, QHeaderView::Stretch);
    header->resizeSection(2, 160);
    header->resizeSection(3, 100);
    layout->addWidget(table_);
}

void QcIssueTable::update_state(const std::vector<domain::Json>& reports) {
    table_->setRowCount(0);
    spatial_by_rule_.clear();
    if (reports.empty()) {
        return;
    }
    const domain::Json& report = reports.front();
    const domain::Json empty_issues = domain::Json::array();
    const domain::Json* issues = &empty_issues;
    const auto iit = report.find("issues");
    if (iit != report.end() && iit->is_array()) {
        issues = &*iit;
    }
    const domain::Json* rules = nullptr;
    const auto rit = report.find("rules");
    if (rit != report.end() && rit->is_array()) {
        rules = &*rit;
    }
    if (rules == nullptr) {
        return;
    }
    spatial_by_rule_ = spatial_issues_by_rule(*issues);
    const auto rows = qc_issue_rows(*rules, *issues);
    table_->setRowCount(int(rows.size()));
    for (int row = 0; row < int(rows.size()); ++row) {
        const auto& spec = rows[std::size_t(row)];
        const QString texts[4] = {
            QString::fromStdString(spec.rule),
            QString::fromStdString(spec.description),
            QString::fromStdString(spec.result_text),
            QString::fromStdString(spec.location),
        };
        for (int col = 0; col < 4; ++col) {
            auto* item = new QTableWidgetItem(texts[col]);
            if (col == 2) {
                item->setForeground(
                    QColor(QString::fromStdString(spec.result_color)));
            }
            item->setFlags(item->flags() & ~Qt::ItemIsEditable);
            table_->setItem(row, col, item);
        }
        table_->setRowHeight(row, 28);
    }
}

std::vector<domain::Json>
QcIssueTable::spatial_issues_for_rule(const std::string& rule) const {
    const auto it = spatial_by_rule_.find(rule);
    return it != spatial_by_rule_.end() ? it->second
                                        : std::vector<domain::Json>{};
}

}  // namespace pwb::ui_review::qt
